File: src/KthNode.cpp
/*
struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};
*/
class Solution {
public:

    vector<TreeNode*> v;

    TreeNode* KthNode(TreeNode* pRoot, unsigned int k)
    {
        if (pRoot == NULL || k < 1) {
            return NULL;
        }
        // 中序遍历
        inOrder(pRoot);
        if (k > v.size()) {
            return NULL;
        } else {
            return v[k-1];
        }
    }

    void inOrder(TreeNode* root) {
        if (root == NULL) return;
        if (root->left) inOrder(root->left);
        v.push_back(root);
        if (root->right) inOrder(root->right);
    }

};

File: src/java/KthNode.java
/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
import java.util.Stack;

public class Solution {

    TreeNode KthNode(TreeNode pRoot, int k)
    {
        if (pRoot == null) {
            return null;
        }

        int count = 0;
        Stack<TreeNode> stack = new Stack<>();
        while (pRoot != null || !stack.empty()) {
            if (pRoot != null) {
                stack.push(pRoot);
                pRoot = pRoot.left;
            } else {
                pRoot = stack.pop();
                if (k == count + 1) {
                    return pRoot;
                } else {
                    count++;
                }
                pRoot = pRoot.right;
            }
        }

        return null;
    }

}
